### PR TITLE
Workarounds for GUI tests running on MacOS

### PR DESCRIPTION
**Pull request by ichebyki**

### CODE REVIEW

**Comment by yalgaer** [comment]
What is the statistics before and after this PR on Mac?

File: src/libs_3rdparty/QSpec/src/utils/GTUtilsDialog.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <core/CustomScenario.h>
 
 #include <QDialogButtonBox>
+#include <QProcess>
```
Comment (yalgaer) [inline src/libs_3rdparty/QSpec/src/utils/GTUtilsDialog.h]: Looks like we do not need this include here.

File: src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+class HI_EXPORT GTUtilsMac {
+
+public:
+    GTUtilsMac() {
+    }
+
+    ~GTUtilsMac() {
+        if (process != nullptr) {
+            process->kill();
```
Comment (yalgaer) [inline src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h]: delete process?

File: src/plugins/GUITestBase/src/tests/PosteriorActions.cpp [hunk 3]
```diff
@@ -89,17 +75,27 @@ POSTERIOR_ACTION_DEFINITION(post_action_0001) {
...
+    fakeClock.startWorkaroundForMacCGEvents(16000, false);
+#endif
         GTWidget::close(os, popupWidget);
         popupWidget = QApplication::activePopupWidget();
     }
 
     QWidget *modalWidget = QApplication::activeModalWidget();
     while (modalWidget != NULL) {
+
+#ifdef Q_OS_MAC
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/PosteriorActions.cpp:94]: nit: code ident looks inconsistent for Q_OS_MAC blocks

File: src/plugins/GUITestBase/src/tests/common_scenarios/annotations/qualifiers/GTTestsAnnotationsQualifiers.cpp [hunk 4]
```diff
@@ -76,7 +76,7 @@ GUI_TEST_CLASS_DEFINITION(test_0001_1) {
     GTFile::copy(os, testDir + "_common_data/scenarios/project/proj2.uprj", sandBoxDir + "proj2.uprj");
     GTFile::copy(os, testDir + "_common_data/scenarios/project/1.gb", sandBoxDir + "1.gb");
 
-    GTFileDialog::openFile(os, testDir + "_common_data/scenarios/project/", "proj2.uprj");
+    GTFileDialog::openFile(os, sandBoxDir, "proj2.uprj");
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/common_scenarios/annotations/qualifiers/GTTestsAnnotationsQualifiers.cpp:79]: Good find!

File: src/plugins/GUITestBase/src/tests/PosteriorActions.cpp (hunk 3 above)
Comment (ichebyki) [reply src/plugins/GUITestBase/src/tests/PosteriorActions.cpp:94]: fixed

File: src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h (hunk 2 above)
Comment (ichebyki) [reply src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h]: yes, this kills the external faked process (this is a root of the workaround)

File: src/libs_3rdparty/QSpec/src/utils/GTUtilsDialog.h (hunk 1 above)
Comment (ichebyki) [reply src/libs_3rdparty/QSpec/src/utils/GTUtilsDialog.h]: fixed

File: src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h (hunk 2 above)
Comment (yalgaer) [reply src/libs_3rdparty/QSpec/src/utils/GTUtilsMac.h]: You allocate a new variable "process" as a part of the class and never delete it.
Could you add 'delete' for the process or use some kind of the shared pointer to clean up memory correctly?

**Review by yalgaer — CHANGES_REQUESTED** [review]

**Comment by ichebyki** [comment]
Fixed https://github.com/ugeneunipro/ugene/pull/202#discussion_r564814747

**Comment by DmitriiSukhomlinov** [comment]
The main question is - have it helped? macOS have been improved?

File: src/libs_3rdparty/QSpec/src/primitives/GTWidget.cpp [hunk 5]
```diff
@@ -40,8 +41,17 @@ namespace HI {
...
+#ifdef Q_OS_MAC
+    GTUtilsMac fakeClock;
+    fakeClock.startWorkaroundForMacCGEvents(16000, false);
+#endif
+
     if (p.isNull()) {
-        p = widget->rect().center();
+        QRect rect = widget->rect();
+        p = rect.center();
+        if (qobject_cast<QLineEdit*>(widget) != nullptr) {
```
Comment (DmitriiSukhomlinov) [inline src/libs_3rdparty/QSpec/src/primitives/GTWidget.cpp:55]: It's not clear what's going on. A comment is required

Comment (ichebyki) [reply src/libs_3rdparty/QSpec/src/primitives/GTWidget.cpp:55]: sure
added #ifdef + comments

**Comment by ichebyki** [comment]
> The main question is - have it helped? macOS have been improved?

yes

**Review by yalgaer — DISMISSED** [review]

**Review by yalgaer — APPROVED** [review]

**Pull request merged**